lib/tests/test_revset.rs

```diff
@@ -2894,6 +2899,134 @@ fn test_evaluate_expression_committer() {
     );
 }
 
+#[test]
+fn test_evaluate_expression_at_operation() {
+    let settings = testutils::user_settings();
+    let test_repo = TestRepo::init();
+    let repo0 = &test_repo.repo;
+    let root_commit = repo0.store().root_commit();
+
+    let mut tx = repo0.start_transaction(&settings);
+    let commit1_op1 = create_random_commit(tx.repo_mut(), &settings)
+        .set_description("commit1@op1")
+        .write()
+        .unwrap();
+    let commit2_op1 = create_random_commit(tx.repo_mut(), &settings)
+        .set_description("commit2@op1")
+        .write()
+        .unwrap();
+    tx.repo_mut()
+        .set_local_bookmark_target("commit1_ref", RefTarget::normal(commit1_op1.id().clone()));
+    let repo1 = tx.commit("test");
+
+    let mut tx = repo1.start_transaction(&settings);
+    let commit1_op2 = tx
+        .repo_mut()
+        .rewrite_commit(&settings, &commit1_op1)
+        .set_description("commit1@op2")
+        .write()
+        .unwrap();
+    let commit3_op2 = create_random_commit(tx.repo_mut(), &settings)
+        .set_description("commit3@op2")
+        .write()
+        .unwrap();
+    tx.repo_mut().rebase_descendants(&settings).unwrap();
+    let repo2 = tx.commit("test");
+
+    let mut tx = repo2.start_transaction(&settings);
+    let _commit4_op3 = create_random_commit(tx.repo_mut(), &settings)
+        .set_description("commit4@op3")
+        .write()
+        .unwrap();
+
+    // Symbol resolution:
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@, commit1_ref)"),
+        vec![commit1_op2.id().clone()]
+    );
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@-, commit1_ref)"),
+        vec![commit1_op1.id().clone()]
+    );
+    assert_matches!(
+        try_resolve_commit_ids(repo2.as_ref(), "at_operation(@--, commit1_ref)"),
+        Err(RevsetResolutionError::NoSuchRevision { .. })
+    );
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "present(at_operation(@--, commit1_ref))"),
+        vec![]
+    );
+
+    // Visibility resolution:
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@, all())"),
+        vec![
+            commit3_op2.id().clone(),
+            commit1_op2.id().clone(),
+            commit2_op1.id().clone(),
+            root_commit.id().clone(),
+        ]
+    );
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@-, all())"),
+        vec![
+            commit2_op1.id().clone(),
+            commit1_op1.id().clone(),
+            root_commit.id().clone(),
+        ]
+    );
+
+    // Operation is resolved relative to the outer ReadonlyRepo.
+    assert_eq!(
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@-, at_operation(@-, all()))"),
+        resolve_commit_ids(repo2.as_ref(), "at_operation(@--, all())")
+    );
+
+    // It's unclear whether "@" should point to the uncommitted operation or
+    // not, but it's resolved to the base operation right now.
```

Comment:
It seems to me that `at_operation(@, x)` should be equivalent to `x` as far as possible

Reply:
I tend to agree, but I said "it's unclear" because a `MutableRepo` doesn't represent an `Operation`. Implementation-wise, it's not easy to resolve `@` (or `@-+`) to a `MutableRepo` because of that. In practice, it's unlikely to evaluate `at_operation()` against `MutableRepo`.

Reply:
Sorry, i didn't mean that as blocking.